### PR TITLE
executor: support Chunk for TableDualExec

### PR DESCRIPTION
**Pull request by zz-jason**
to #5261

### CODE REVIEW

**Comment by coocood** [comment]
numDualRows is always 1.
If not `done`, We can just write one row, set `done` to true and return.

File: executor/builder.go [hunk 1]
```diff
@@ -692,10 +693,20 @@ func (b *executorBuilder) buildProjection(v *plan.PhysicalProjection) Executor {
...
 func (b *executorBuilder) buildTableDual(v *plan.PhysicalTableDual) Executor {
-	return &TableDualExec{
-		baseExecutor: newBaseExecutor(v.Schema(), b.ctx),
-		rowCount:     v.RowCount,
+	if v.RowCount != 0 && v.RowCount != 1 {
+		b.err = errors.Errorf("buildTableDual failed, invalid row count for dual table: %v", v.RowCount)
+		return nil
+	}
+	dualSchema := v.Schema()
+	if dualSchema.Len() == 0 {
```
Comment (shenli) [inline executor/builder.go]: Why change the schema?

Comment (zz-jason) [reply executor/builder.go]: the schema can not be empty, we need to use it to build a chunk for `TableDualExec` and append some value into it to indicate there exists some rows in that chunk.

Comment (zz-jason) [reply executor/builder.go]: this work should be done in plan building stage

**Comment by zz-jason** [comment]
@hanfei1991 @coocood PTAL

**Comment by coocood** [comment]
@zz-jason 
I think a better way to solve the problem is to define a `noColumnLen` in `Chunk`, then when we don't have any columns, we use this value for `NumRows()`.

**Comment by shenli** [comment]
Please resolve the conflicts.

File: executor/adapter.go [hunk 2]
```diff
@@ -228,9 +228,11 @@ func (a *ExecStmt) Exec(goCtx goctx.Context) (ast.RecordSet, error) {
 		// Update processinfo, ShowProcess() will use it.
 		pi.SetProcessInfo(sql)
 	}
-	// Fields or Schema are only used for statements that return result set.
+	// "e.Schema().Len() == 0" means the executor doesn't return any result to the client, we execute it without delay.
 	if e.Schema().Len() == 0 {
 		return a.handleNoDelayExecutor(goCtx, e, ctx, pi)
+	} else if proj, ok := e.(*ProjectionExec); ok && proj.calculateNoDelay {
```
Comment (shenli) [inline executor/adapter.go]: Please comment this condition branch. Why we need this?

File: plan/planbuilder.go [hunk 3]
```diff
@@ -257,7 +270,11 @@ func (b *planBuilder) buildSet(v *ast.SetStmt) Plan {
 				vars.Value = ast.NewValueExpr(cn.Name.Name.O)
 			}
 			mockTablePlan := LogicalTableDual{}.init(b.ctx)
-			mockTablePlan.SetSchema(expression.NewSchema())
+			mockTablePlan.SetSchema(expression.NewSchema(&expression.Column{
```
Comment (shenli) [inline plan/planbuilder.go]: Why this is changed.

**Comment by coocood** [comment]
LGTM

File: util/chunk/chunk.go [hunk 4]
```diff
@@ -174,6 +186,7 @@ func (c *Chunk) Append(other *Chunk, begin, end int) {
 			dst.length++
 		}
 	}
+	c.numVirtualRows += other.numVirtualRows
```
Comment (winkyao) [inline util/chunk/chunk.go]: why add all other.numVirtualRows? if we call Append more than once

Comment (zz-jason) [reply util/chunk/chunk.go]: a mistake, fixed

File: plan/planbuilder.go [hunk 5]
```diff
@@ -225,20 +225,29 @@ func (b *planBuilder) buildExecute(v *ast.ExecuteStmt) Plan {
...
+			FromID:   p.id,
+			Position: schema.Len() + 1,
+			RetType:  expr.GetType(),
+		})
 	}
-	dual.SetSchema(expression.NewSchema())
-	p := LogicalProjection{Exprs: exprs}.init(b.ctx)
 	p.SetChildren(dual)
+	p.self = p
 	p.SetSchema(expression.NewSchema())
```
Comment (zimulala) [inline plan/planbuilder.go]: Could we remove this line?

Comment (zz-jason) [reply plan/planbuilder.go]: removed

**Pull request merged**